src/box/lua/iproto.c

```diff
@@ -268,16 +299,124 @@ push_iproto_protocol_features(struct lua_State *L)
 					  lengthof(features));
 }
 
+/**
+ * Encodes Lua table packet part to MsgPack using IPROTO key translation table:
+ * uses @a ibuf to store the result and saves the result length in @a mp_len.
+ */
+static const char *
+encode_lua_table_packet_part_to_mp(struct lua_State *L, int idx, size_t *mp_len)
+{
+	struct region *gc = &fiber()->gc;
+	struct mpstream stream;
+	mpstream_init(&stream, gc, region_reserve_cb, region_alloc_cb,
+		      luamp_error, L);
+	size_t used = region_used(gc);
+	luamp_encode_with_translation(L, luaL_msgpack_default, &stream, idx,
+				      iproto_key_translation);
+	mpstream_flush(&stream);
+	*mp_len = region_used(gc) - used;
+
+	const char *mp = region_join(gc, *mp_len);
+	if (mp == NULL) {
+		diag_set(OutOfMemory, *mp_len, "mp", "MsgPack data");
+		region_truncate(gc, used);
+	}
+	return mp;
+}
+
+/**
+ * Converts a packet part argument to MsgPack: if the argument is a string, then
+ * no conversion is needed — otherwise the argument must be a Lua table, which
+ * is encoded using `encode_lua_table_packet_part_to_mp`.
+ */
+static const char *
+convert_packet_part_arg_to_mp(struct lua_State *L, int idx, size_t *mp_len)
+{
+	int packet_part_type = lua_type(L, idx);
+	if (packet_part_type != LUA_TSTRING && packet_part_type != LUA_TTABLE) {
+		const char *arg = tt_sprintf("expected %d argument to be a "
+					     "string or table", idx);
+		diag_set(LuajitError, arg);
+		return NULL;
+	}
+	return (packet_part_type == LUA_TSTRING) ?
+	       lua_tolstring(L, idx, mp_len) :
+	       encode_lua_table_packet_part_to_mp(L, idx, mp_len);
+}
+
+/**
+ * Lua interface for `box_iproto_send`.
+ */
+static int
+lbox_iproto_send(struct lua_State *L)
+{
+	int n_args = lua_gettop(L);
+	if (n_args < 2 || n_args > 3)
+		return luaL_error(L, "Usage: "
+				     "box.iproto.send(sid, header<, body>)");
+	uint64_t sid = luaL_touint64(L, 1);
+
+	size_t used = box_region_used();
+
+	size_t header_len;
+	const char *header =
+		convert_packet_part_arg_to_mp(L, 2, &header_len);
+	if (header == NULL)
+		return luaT_error(L);
+
+	size_t body_len = 0;
+	const char *body = NULL;
+	if (lua_gettop(L) == 3) {
+		body = convert_packet_part_arg_to_mp(L, 3, &body_len);
+		if (body == NULL) {
+			box_region_truncate(used);
+			return luaT_error(L);
+		}
+	}
+
+	int rc = box_iproto_send(sid,
+				 header, header + header_len,
+				 body, body + body_len);
+	box_region_truncate(used);
+	return (rc == 0) ? 1 : luaT_error(L);
```

Comment:
What does it return on success?

Reply:
Nothing.

Reply:
Returning 1 means it pushed one value to Lua stack to be returned. What is it?

Reply:
I didn't know that the return value of C functions called from Lua is the number of return values, fixed.